### PR TITLE
coverage: Simplify creation of sum counters

### PR DESCRIPTION
**Pull request by Zalathar**
A small and self-contained improvement, extracted from some larger changes that I'm still working on.

Ultimately I want to avoid creating these sum counter-expressions in some cases (in favour of just adding physical counters directly to the nodes we care about), so a good incremental move towards that is splitting the “gather edge counters” step out from the ”build a sum of those counters” step.

Creating an extra intermediate vector should have negligible cost (and coverage isn't exercised by the benchmark suite anyway). The removed logging is redundant with the `#[instrument(..)]` logging we already have on the underlying method calls.

### CODE REVIEW

File: compiler/rustc_mir_transform/src/coverage/counters.rs [hunk 1]
```diff
@@ -155,12 +155,14 @@ impl CoverageCounters {
...
     ///
-    /// This is useful when using [`Iterator::fold`] to build an arbitrary-length sum.
-    fn make_sum_expression(&mut self, lhs: Option<BcbCounter>, rhs: BcbCounter) -> BcbCounter {
-        let Some(lhs) = lhs else { return rhs };
-        self.make_expression(lhs, Op::Add, rhs)
+    /// Returns `None` if the given list of counters was empty.
+    fn make_sum(&mut self, counters: &[BcbCounter]) -> Option<BcbCounter> {
+        counters
+            .iter()
+            .copied()
```
Comment (Swatinem) [inline compiler/rustc_mir_transform/src/coverage/counters.rs:164]:
```suggestion
    fn make_sum(&mut self, counters: impl Iterator<Item = BcbCounter>) -> Option<BcbCounter> {
        counters
```

File: compiler/rustc_mir_transform/src/coverage/counters.rs [hunk 2]
```diff
@@ -315,20 +317,17 @@ impl<'a> MakeBcbCounters<'a> {
...
+        let other_out_edge_counters = successors
+            .iter()
+            .copied()
+            // Skip the chosen edge, since we'll calculate its count from this sum.
+            .filter(|&to_bcb| to_bcb != expression_to_bcb)
+            .map(|to_bcb| self.get_or_make_edge_counter(from_bcb, to_bcb))
+            .collect::<Vec<_>>();
+        let sum_of_all_other_out_edges: BcbCounter = self
+            .coverage_counters
+            .make_sum(&other_out_edge_counters)
```
Comment (Swatinem) [inline compiler/rustc_mir_transform/src/coverage/counters.rs:329]:
```suggestion
            .map(|to_bcb| self.get_or_make_edge_counter(from_bcb, to_bcb));
        let sum_of_all_other_out_edges: BcbCounter = self
            .coverage_counters
            .make_sum(other_out_edge_counters)
```

same down below

**Review by Swatinem — COMMENTED** [review]
I would suggest taking an iterator instead of forcing a `collect`. Unless you really need this to be a slice for some followup work. but even then, you could move the `.iter().cloned()` to the caller.

**Comment by Swatinem** [comment]
I guess I should have read the description more carefully 🙈 don’t mind me then, as it looks like you will be needing a `Vec` in the future here anyway.

**Comment by Zalathar** [comment]
Yeah, creating the intermediate vector is deliberate, so that the two steps are clearly separate (and can be pulled further apart without additional hassle).

**Comment by Zalathar** [comment]
And while I *could* make the helper take an iterator instead of a slice, I prefer the direct simplicity of the slice.

It would be easy enough to change that back to taking an iterator if necessary, but I don't foresee that happening in the near future.

**Comment by Zalathar** [comment]
Hmm, was rustbot asleep?

r? compiler

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**